orttraining/orttraining/training_api/checkpoint.cc

```diff
@@ -194,43 +201,108 @@ Status FromTensorProtos(
 
   constexpr size_t m_bytes = 1024 * 1024;
   size_t fbs_buffer_size = 0U;
+  size_t fbs_potential_external_buffer_size = 0;
   for (const auto& tensor_proto : trainable_tensor_protos) {
-    fbs_buffer_size += tensor_proto.ByteSizeLong();
+    auto bytes = tensor_proto.ByteSizeLong();
+    fbs_buffer_size += bytes;
+
+    if (bytes > onnxruntime::fbs::utils::kMinimumSizeForExternalData) {  // assuming strings aren't trainable
+      fbs_potential_external_buffer_size += bytes;
+    }
   }
+
   for (const auto& tensor_proto : non_trainable_tensor_protos) {
-    fbs_buffer_size += tensor_proto.ByteSizeLong();
+    auto bytes = tensor_proto.ByteSizeLong();
+    fbs_buffer_size += bytes;
+
+    if (bytes > onnxruntime::fbs::utils::kMinimumSizeForExternalData &&
+        tensor_proto.data_type() != ONNX_NAMESPACE::TensorProto_DataType_STRING) {
+      fbs_potential_external_buffer_size += bytes;
+    }
   }
 
   // Align buffer size to 1MB.
+  // TODO: This should probably also have an allowance for the other parts of the tensor if we're trying to avoid
+  // all reallocs during serialization. e.g. name, shape, data type, offset if external data is used.
   fbs_buffer_size = std::max(fbs_buffer_size, m_bytes);
   fbs_buffer_size = ((fbs_buffer_size + m_bytes - 1) / m_bytes) * m_bytes;
+
+  // 2GB is the real limit, but fbs_buffer_size doesn't include space for initializer names and shapes.
+  // for now, arbitrarily use ~1.8GB as a conservative 'safe' value with a 10% buffer.
+  const bool use_external_data = fbs_buffer_size >= 1800 * m_bytes;
+
+  fbs::utils::ExternalDataWriter external_data_writer = nullptr;
+  std::optional<std::ofstream> external_data_stream;
+
+  if (use_external_data) {
+    // 2GB. Have to use external file for data.
+    fbs_buffer_size -= fbs_potential_external_buffer_size;  // reduce fbs buffer size to account for external data
+
+    auto data_path = ExternalCheckpointDataPath(checkpoint_path);
+    external_data_stream = std::ofstream(data_path, std::ios::binary);
+
+    ORT_RETURN_IF(external_data_stream->fail(), "Failed to create checkpoint's external data file: ",
+                  ToUTF8String(data_path));
+
+    // setup the data writer to write aligned data to external_data_stream
```

Comment:
Considering that the data is already available in an external file since ONNX models also pose a 2 GB size constraint, shall we avoid creating another external file and re-use the original one.
We can add code to read from that external file (on the device) and avoid creating a duplicate.

Reply:
I'm not very familiar with the system as a whole. If there's already an external file that we could use that's great. Would we re-use the offset from the original TensorProto and maybe have a requirement that only one external file is involved (otherwise we also need to store a filename in the flatbuffer data)?

Reply:
For visibility, we decided on generating a single checkpoint external data file in the generate_artifacts step, since ONNX models (depending on export method) will generate a separate external data file for every weight, which would not be ideal for on-device training.